src/rbindlist.c

```diff
@@ -514,9 +514,33 @@ SEXP rbindlist(SEXP l, SEXP usenamesArg, SEXP fillArg, SEXP idcolArg)
         if (w==-1 || !length(thisCol=VECTOR_ELT(li, w))) {  // !length for zeroCol warning above; #1871
           writeNA(target, ansloc, thisnrow);  // writeNA is integer64 aware and writes INT64_MIN
         } else {
-          if ((TYPEOF(target)==VECSXP || TYPEOF(target)==EXPRSXP) && TYPEOF(thisCol)!=TYPEOF(target)) {
+          if ((TYPEOF(target)==VECSXP) && (isVectorAtomic(thisCol) || TYPEOF(thisCol)==LISTSXP)) {
             // do an as.list() on the atomic column; #3528
+            // pairlists (LISTSXP) can also be coerced to lists using coerceVector
             thisCol = PROTECT(coerceVector(thisCol, TYPEOF(target))); nprotect++;
+          } else if ((TYPEOF(target)==VECSXP) && TYPEOF(thisCol)==EXPRSXP) {
+            // For EXPRSXP each element to be wrapped in a list, e.g. expression vectors #546
+            SEXP thisColList = PROTECT(allocVector(VECSXP, length(thisCol))); nprotect++;
+            for(int r=0; r<length(thisCol); ++r) {
+              SEXP thisElement = VECTOR_ELT(thisCol, r);
+              thisElement = PROTECT(coerceVector(thisElement, EXPRSXP)); nprotect++; // otherwise LANGSXP
+              SET_VECTOR_ELT(thisColList, r, thisElement);
+            }
+            thisCol = thisColList;
+          } else if ((TYPEOF(target)==VECSXP) && !isVector(thisCol) && TYPEOF(thisCol)!=TYPEOF(target)) { 
+            // Anything not a vector we can assign directly through SET_VECTOR_ELT
+            // Although tecnically there should only be one list element for any type met here,
+            // the length of the type may be > 1, in which case the other columns in data.table
+            // will have been recycled. We therefore in turn have to recycle the list elements
+            // to match the number of rows.
+            SEXP thisColList = PROTECT(allocVector(VECSXP, length(thisCol))); nprotect++;
+            for(int r=0; r<length(thisCol); ++r) {
+              SET_VECTOR_ELT(thisColList, r, thisCol);
+            }
```

Comment:
Demonstration of the recycling issue and why we need this for loop:

```
> A = data.table(a=0L, B=quote(1+1))
> A
       a      B
   <int> <call>
1:     0  1 + 1
2:     0  1 + 1
3:     0  1 + 1
```

The for loop means we get the right result with rbind:

```
> rbind(A)
       a         B
   <int>    <list>
1:     0 <call[3]>
2:     0 <call[3]>
3:     0 <call[3]>
```

Otherwise we would have

```
> rbind(A)
       a         B
   <int>    <list>
1:     0 <call[3]>
2:     0 
3:     0
```

It would be good to fix this in data.table()/as.data.table() so that these non-vector types are wrapped in list at the time of data.table construction

Reply:
agree, very nice demo